### PR TITLE
Search package: run auto config from activation API

### LINKED ISSUE #22588: Search: do not add additional Search widget to sidebar if the sidebar already includes a Search widget
This applies to non-block themes, supporting at least one widget area. Such themes, when you launch the site, will add a set of default widgets to the first sidebar area. Among those default widgets you will find the Search widget.

Jetpack can then add its own Search widget to the sidebar when you first purchase a Search plan. Ideally, we would want to avoid adding an extra Search widget when the site already has one set up.

*****

Reported here:
p8oabR-Nm-p2#comment-5986

### PR DESCRIPTION
**Pull request by kangzj**
#### Changes proposed in this Pull Request:
The PR tries to make the search auto config process idempotent and calls auto config from search activation API. The PR needs `D71473-code` to be functional however it could be shipped without it. Related: #22084. Also fixes #22588.

- Refactored `auto_config_search` functionality to be idempotent and added tests
- Added options NOT to run activation or auto config for the Activation API (changed payload format the activation API consumes, see smoking test section)
- Fixed an issue where `Classic_Search` and `Instant_Search` share the same static `$instance` variable

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Go to a site with Jetpack Search subscription
* Use a non block theme (Block themes are supported thru #22613 )
* Enable Instant Search
* Remove search widgets from sidebars
* Remove overlay sidebar widgets
* Run `jetpack docker wp jetpack-search auto_config username/id`
* Ensure the removed widgets are added again with default filters

* Ensure PHP tests pass: `( cd projects/packages/search && composer test-php )`

##### Smoke test activate plan API:
Note: 
- Remove all widgets in theme sidebar and ovleray sidebar
- All keys in the JSON payload are optional.
- cookie and nonce could be obtained from page `/wp-admin/admin.php?page=jetpack-search`, inspect API requests like `/wp-json/jetpack/v4/search/settings?_cacheBuster=1644280717075`

```
POST https://wordpress.local/?rest_route=/jetpack/v4/search/plan/activate HTTP/1.1
Content-Type: application/json
x-wp-nonce: {{nonce}}
cookie: {{cookie}}

{
        "auto_config_search": false,// default is true,
        "enable_search": true, // default is true,
        "enable_instant_search": true, // default is true,
	"search_plan_info": {
            "search_subscriptions": [{
		"ID": "1",
		"user_id": "1",
		"blog_id": "1",
		"product_id": "2001",
		"expiry": "2117-06-30",
		"subscribed_date": "2017-06-30 16:07:06",
		"renew": true,
		"auto_renew": true,
		"ownership_id": "9698106",
		"most_recent_renew_date": "",
		"subscription_status": "active",
		"product_name": "Jetpack Professional",
		"product_name_en": "Jetpack Professional",
		"product_slug": "jetpack_business",
		"product_type": "bundle",
		"cost": 0,
		"currency": "USD",
		"bill_period": "365",
		"available": "yes",
		"multi": false,
		"support_document": null,
		"is_instant_search": false,
		"tier": null
	}],
	"supports_instant_search": true,
	"supports_only_classic_search": false,
	"supports_search": true,
	"default_upgrade_bill_period": "yearly"
    }
}

```

### CODE REVIEW

**Review by jsnmoon — COMMENTED** [review]
Excellent refactor! Our auto-config code path was in severe need of a clean-up, and this PR provides that in droves 👏 👏 👏 

I left some documentation suggestions and questions about the test code.

Since the PR stated it should be functional without D71473-code, I tried smoke testing these changes without the code patch. I ran into some strange bugs following the testing instructions; here are my repro steps:

1. Switched to a non-block theme (2020).
2. Deleted all widgets from all sidebars via `/wp-admin/widgets.php`.
3. Triggered auto-config via `jetpack docker wp jetpack-search auto_config`.
4. Noticed two erroneous results:
    1.  No Jetpack Search widgets were added to the search overlay sidebar.
    2. Two Jetpack Search widgets were added to Footer 1 sidebar.
Screen Shot 2022-02-09 at 4 20 15 PM
5. This also made the two Jetpack Search widgets impossible to remove via the new block-based widget editor. Doing so summons a cryptic error toast on the bottom left of the page.

Screen Shot 2022-02-09 at 4 22 44 PM

Here is the entirety of my test site, exported into XML (and compressed into ZIP for uploading to Github): [jason039stestsite.WordPress.2022-02-09.xml.zip](https://github.com/Automattic/jetpack/files/8036818/jason039stestsite.WordPress.2022-02-09.xml.zip)

File: projects/packages/search/src/instant-search/class-instant-search.php [hunk 1]
```diff
@@ -309,99 +309,124 @@ public function get_search_aggregations_results() {
...
+	public function get_jp_search_wediget_next_id() {
+		$widget_opt_name = Helper::get_widget_option_name();
+		$widget_options  = get_option( $widget_opt_name, array() );
+		return ! empty( $widget_options ) ? intval( max( array_keys( $widget_options ) ) ) + 1 : 1;
+	}
+
+	/**
+	 * Get search and JP Search widget indices in theme sidebar.
+	 */
+	protected function get_search_widget_indices() {
```
[the remark this answers is not in the document]
Comment (kangzj) [reply projects/packages/search/src/instant-search/class-instant-search.php:410]: Good point which will make the function more reusable 👍

File: projects/packages/search/src/instant-search/class-instant-search.php [hunk 2]
```diff
@@ -509,21 +535,30 @@ public function auto_config_excluded_post_types() {
...
 		// Check if WooCommerce plugin is active (based on https://docs.woocommerce.com/document/create-a-plugin/).
-		if ( ! in_array(
+		if ( in_array(
 			'woocommerce/woocommerce.php',
 			apply_filters( 'active_plugins', Helper::get_active_plugins() ),
 			true
 		) ) {
-			return false;
+			update_option( $result_format_option_name, Options::RESULT_FORMAT_PRODUCT );
+			return;
```
[the remark this answers is not in the document]
Comment (kangzj) [reply projects/packages/search/src/instant-search/class-instant-search.php:557]: `$result_format_option_name` is used three times in the function and that's the reason it's extracted to a variable 😄

[the remark this answers is not in the document]
Comment (kangzj) [reply projects/packages/search/src/instant-search/class-instant-search.php:557]: could you clarify on this one please? Sorry I don't quite get it.

**Comment by kangzj** [comment]
> I ran into some strange bugs following the testing instructions; 

Thanks for catching it. There was an bug - a mis-use of id / index. Now fixed.

Also chunked in a fix for #22588.

File: projects/packages/search/src/instant-search/class-instant-search.php [hunk 3]
```diff
@@ -309,97 +311,161 @@ public function get_search_aggregations_results() {
...
+		}
+		// test search block widget.
+		if ( 0 === strpos( $widget_id, 'block-' ) ) {
+			$widget_blocks = get_option( 'widget_block', array() );
+			$widget_index  = str_replace( 'block-', '', $widget_id );
+			// A single block could be of type string or array.
+			if ( isset( $widget_blocks[ $widget_index ]['content'] ) && false !== strpos( (string) $widget_blocks[ $widget_index ]['content'], 'wp:search' ) ) {
+				return true;
+			}
+			if ( isset( $widget_blocks[ $widget_index ] ) && false !== strpos( (string) $widget_blocks[ $widget_index ], 'wp:search' ) ) {
```
Comment (jsnmoon) [inline projects/packages/search/src/instant-search/class-instant-search.php:464]: I'm getting an `Array to string conversion` warning here in my errors log.

**Review by jsnmoon — COMMENTED** [review]
Usual flows appear to work as expected (e.g. search module disabled & search module + instant search enabled), but executing the activation on a site with only classic search enabled throws a 500 ISE.

File: projects/packages/search/src/instant-search/class-instant-search.php [hunk 4]
```diff
@@ -309,97 +311,161 @@ public function get_search_aggregations_results() {
...
-				// Add to top.
-				array_unshift( $sidebars['sidebar-1'], Helper::build_widget_id( $next_id ) );
+		// If there's JP search widget in theme sidebar, skip.
+		foreach ( (array) $sidebars[ self::AUTO_CONFIG_SIDEBAR ] as $widget_id ) {
+			if ( 0 === strpos( $widget_id, Helper::FILTER_WIDGET_BASE ) ) {
+				// Already configured.
+				return;
 			}
-			$next_id++;
 		}
```
Comment (jsnmoon) [inline projects/packages/search/src/instant-search/class-instant-search.php:379]: Like with `auto_config_overlay_sidebar_widgets`, can we replace this by checking if the second value of `$this->get_search_widget_indices( $sidebars )` (L384) is false?

File: projects/packages/search/src/instant-search/class-instant-search.php [hunk 5]
```diff
@@ -309,97 +311,161 @@ public function get_search_aggregations_results() {
...
+		// If there's JP search widget in overly sidebar, skip.
 		if ( isset( $sidebars['jetpack-instant-search-sidebar'] ) ) {
 			foreach ( (array) $sidebars['jetpack-instant-search-sidebar'] as $widget_id ) {
-				if ( 0 === strpos( $widget_id, $slug ) ) {
+				if ( 0 === strpos( $widget_id, Helper::FILTER_WIDGET_BASE ) ) {
 					// Already configured.
 					return;
 				}
 			}
 		}
```
Comment (jsnmoon) [inline projects/packages/search/src/instant-search/class-instant-search.php:337]: Could we replace this by checking if `$sidebar_jp_searchbox_wiget_id` (L349) is false?

**Review by jsnmoon — COMMENTED** [review]
Changes look reasonable! This looks good to ship once these two items are resolved:

- [x] ISE is thrown if auto-configuring a site with only classic search enabled
- [x] Warning notice on L464 is logged during auto-configuration.

**Comment by kangzj** [comment]
HI @jsnmoon thanks for the review.

>  ISE is thrown if auto-configuring a site with only classic search enabled

It's resolved by 2b02e9a76b where `Instant_Search` shouldn't share singleton with `Classic_Search`.

**Comment by kangzj** [comment]
>  Warning notice on L464 is logged during auto-configuration.

Addressed in afefb4eaf5

File: projects/packages/search/src/instant-search/class-instant-search.php (hunk 4 above)
Comment (kangzj) [reply projects/packages/search/src/instant-search/class-instant-search.php]: Addressed in 352d1ad86f

File: projects/packages/search/src/instant-search/class-instant-search.php (hunk 5 above)
Comment (kangzj) [reply projects/packages/search/src/instant-search/class-instant-search.php:345]: Addressed in 3839f610c1

**Comment by kangzj** [comment]
Hi @jsnmoon thanks for your speedy review, especially when you are super busy. All feedback is now addressed. Would you mind taking another look? Thanks.

**Comment by kangzj** [comment]
Hi @Automattic/jetpack-crew lock checks wouldn't pass for this one too.

**Comment by kraftbj** [comment]
Fixed it up with `tools/composer-update-monorepo.sh --root-reqs "projects/plugins/backup"` from the CI error.

**Review by jeherve — APPROVED** [review]
This works well for me. 🚢

**Pull request merged**